ipset.go

```diff
@@ -99,154 +99,135 @@ func (s *IPSet) RemoveSet(b *IPSet) {
 	}
 }
 
-// point is either the start or end of IP range of wanted or unwanted
-// IPs.
-// This is used by the implementation of IPSet.Ranges.
-type point struct {
-	ip    IP
-	want  bool // true for 'add', false for remove
-	start bool // true for start of range, false for (inclusive) end
-}
-
-// Less sorts points by the needs of the IPSet.Ranges function.
-// See also comments in netaddr_test.go's TestPointLess.
-func (a point) Less(b point) bool {
-	cmp := a.ip.Compare(b.ip)
-	if cmp != 0 {
-		return cmp < 0
-	}
-	if a.want != b.want {
-		if a.start == b.start {
-			return !a.want
-		}
-		return a.start
-	}
-	if a.start != b.start {
-		return a.start
-	}
-	return false
-}
-
 func discardf(format string, args ...interface{}) {}
 
 // debugf is reassigned by tests.
 var debugf = discardf
 
-func debugLogPoints(points []point) {
-	for _, p := range points {
-		emo := "✅"
-		if !p.want {
-			emo = "❌"
-		}
-		if p.start {
-			debugf(" {  %-15s %s\n", p.ip, emo)
-		} else {
-			debugf("  } %-15s %s\n", p.ip, emo)
-		}
-	}
-}
-
 // Ranges returns the minimum and sorted set of IP
 // ranges that covers s.
 func (s *IPSet) Ranges() []IPRange {
-	var points []point
-	for _, r := range s.in {
-		points = append(points, point{r.From, true, true}, point{r.To, true, false})
+	const debug = false
+	if debug {
+		debugf("ranges start in=%v out=%v", s.in, s.out)
 	}
-	for _, r := range s.out {
-		points = append(points, point{r.From, false, true}, point{r.To, false, false})
+	in, ok := mergeIPRanges(s.in)
+	if !ok {
+		return nil
+	}
+	out, ok := mergeIPRanges(s.out)
+	if !ok {
+		return nil
 	}
-	sort.Slice(points, func(i, j int) bool { return points[i].Less(points[j]) })
-	const debug = false
 	if debug {
-		debugf("post-sort:")
-		debugLogPoints(points)
-		debugf("merging...")
+		debugf("ranges sort  in=%v out=%v", in, out)
 	}
 
-	// Now build 'want', like points but with "remove" ranges removed
-	// and adjacent blocks merged, and all elements alternating between
-	// start and end.
-	want := points[:0]
-	var addDepth, removeDepth int
-	for i, p := range points {
-		depth := &addDepth
-		if !p.want {
-			depth = &removeDepth
-		}
-		if p.start {
-			*depth++
-		} else {
-			*depth--
-		}
+	// in and out are sorted in ascending range order, and have no
+	// overlaps within each other. We can run a merge of the two lists
+	// in one pass.
+
+	ret := make([]IPRange, 0, len(in))
+	for len(in) > 0 && len(out) > 0 {
```

Comment:
Maybe give in[0] and out[0] names to eliminate some of the syntactical stutter below?
 
```
rin, rout := in[0], out[0]
```

or `inr, outr`